### PR TITLE
Ab#22

### PR DESCRIPTION
**Pull request by AdamPrisen**
This pull request adds support for fetching metadata for contract and token level in our blockchain indexer. It also introduces indexing for an additional collection.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -130,41 +121,41 @@ async function processTransfersERC721(ctx: Context, transfersData: TransferEvent
...
+        }))
+        await new Promise(f => setTimeout(f, 10000));
+
     }
 }
 
 
-export async function getOrCreateNftCollections(ctx: Context, cache: Cache, collectionsData: Map<string, CollectionData>){
+export async function getOrCreateNftCollections(ctx: Context, latestBlockNumber: number, cache: Cache, collectionsData: Map<string, TransferEvent>){
     for (let batch of splitIntoBatches([...collectionsData.keys()])){
```
Comment (brano-hozza) [inline src/main.ts:159]: Again

File: src/main.ts [hunk 2]
```diff
@@ -179,8 +170,71 @@ export async function getOrCreateNftCollections(ctx: Context, cache: Cache, coll
             blockchain: collectionData.blockchain,
             contractType: collectionData.contractType
        });
+       await fillCollectionMetadata(ctx, latestBlockNumber, nftCollenctionEntity)
        cache.NftCollections.set(collectionId, nftCollenctionEntity);
     }
 }
 
+export async function fillTokenMetadata(ctx: Context, latestBlockNumber: number, token: NftEntity){
+    let tokenUri = null
```
Comment (brano-hozza) [inline src/main.ts:179]: This could be done prettier

File: src/main.ts [hunk 3]
```diff
@@ -179,8 +170,71 @@ export async function getOrCreateNftCollections(ctx: Context, cache: Cache, coll
...
+
+}
+
+export async function fillCollectionMetadata(ctx: Context, latestBlockNumber: number, collection: NftCollectionEntity){
+    const contract = new erc721.Contract(ctx, {height: latestBlockNumber}, collection.address);
+    let contractUri = null
+    try{
+        contractUri = await contract.contractURI();
+    } catch{
+        contractUri = null
```
Comment (brano-hozza) [inline src/main.ts:213]: No need to do that

File: src/main.ts [hunk 4]
```diff
@@ -179,8 +170,71 @@ export async function getOrCreateNftCollections(ctx: Context, cache: Cache, coll
...
+    }
+    if (contractUri !== null){
+        collection.uri = contractUri
+        Object.assign(collection, await parseContractMetadata(ctx, contractUri))
+        
+    } else {
+        collection.name = await contract.name()
```
Comment (brano-hozza) [inline src/main.ts:220]: Cant we access all these async data in one call?

File: src/processor.ts [hunk 5]
```diff
@@ -29,10 +35,11 @@ export const processor = new EvmBatchProcessor()
         }
     })
     .addLog({
-        address: [CONTRACT_ADDRESS],
+        address:[...CONTRACT_ADDRESSES.keys()],
```
Comment (brano-hozza) [inline src/processor.ts:38]: Again

File: src/utils/metadata.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    return url
+  }
+
+
+export async function parseContractMetadata(ctx: Context, contractUri: string): Promise<ContractMetadata|null>{
+    try{
+        const { status, data } = await api.get(replaceIpfsUrl(contractUri))
+        if (status < 400) {
+            return data
```
Comment (brano-hozza) [inline src/utils/metadata.ts:46]: Here add some log error

File: src/utils/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        ctx.log.warn(`ContractMetadata ERROR ${contractUri}  ${(e as Error).message}`)
+        return null
+    }
+}
+
+export async function parseTokenMetadata(ctx: Context, tokenUri: string): Promise<TokenMetadata|null>{
+    try{
+        const { status, data } = await api.get(replaceIpfsUrl(tokenUri))
+        if (status < 400) {
+            return data
```
Comment (brano-hozza) [inline src/utils/metadata.ts:59]: Log error again

**Review by brano-hozza — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 8]
```diff
@@ -130,41 +121,41 @@ async function processTransfersERC721(ctx: Context, transfersData: TransferEvent
...
-       const collectionId = getCollectionEntityId(nftData.contractAddress, nftData.blockchain)
-       const nftCollenction = cache.NftCollections.get(collectionId)
-       let nftEntity = new NftEntity({
-            id: nftId,
-            tokenId: nftData.tokenId,
-            nftCollection: nftCollenction
-       });
-       cache.Nfts.set(nftId, nftEntity);
+    await getOrCreateNftCollections(ctx, latestBlockNumber, cache, collectionsData)
+    for (let batch of splitIntoBatches([...nftsData.entries()], 25)){
```
[the remark this answers is not in the document]
Comment (AdamPrisen) [reply src/main.ts:140]: Map.keys(), Map.entries(), Map.values() returns iterator, spread is used to iterate the iterator and convert to array

File: src/main.ts [hunk 9]
```diff
@@ -130,41 +121,41 @@ async function processTransfersERC721(ctx: Context, transfersData: TransferEvent
 }
 
 
-export async function getOrCreateNfts(ctx: Context, cache: Cache, nftsData: Map<string, NftData>){
+export async function getOrCreateNfts(ctx: Context, latestBlockNumber: number, cache: Cache, nftsData: Map<string, TransferEvent>){
     for (let batch of splitIntoBatches([...nftsData.keys()])){
```
[the remark this answers is not in the document]
Comment (AdamPrisen) [reply src/main.ts:125]: spread is used to iterate the iterator and convert to array

File: src/main.ts (hunk 1 above)
Comment (AdamPrisen) [reply src/main.ts:159]: spread is used to iterate the iterator and convert to array

File: src/main.ts (hunk 2 above)
Comment (AdamPrisen) [reply src/main.ts]: i tried 👍

File: src/main.ts (hunk 3 above)
Comment (AdamPrisen) [reply src/main.ts]: fixed

File: src/main.ts (hunk 4 above)
Comment (AdamPrisen) [reply src/main.ts]: No, but for now, i changed it to make this calls at least asynchronously

File: src/utils/metadata.ts (hunk 6 above)
Comment (AdamPrisen) [reply src/utils/metadata.ts:46]: fixed

File: src/utils/metadata.ts (hunk 7 above)
Comment (AdamPrisen) [reply src/utils/metadata.ts:60]: fixed

**Comment by AdamPrisen** [comment]
@brano-hozza could you please take a look at the changes I made, it should be fixed or commented.

**Review by brano-hozza — APPROVED** [review]

**Pull request merged**